[extras] add merge_safe as an alternative to merge_auto

pabloyoyoista opened the pull request:
Sofia has recurrently expressed dissatisfaction with merge_auto not working super well at reproducing data due to the update in some data types not being fit for data science. For example, while Counter's update does what one would expect, dict's update might simply override previous values from the iterator without giving out any feedback. To allow data scientists to avoid this issue, we add a new API that makes sure that previous values don't exist.

This is probably missing tests and some further discussion (like, should we consider the `Counter` type safe?), so marking it as draft

pabloyoyoista committed:
[extras] add allow_override as an option to merge_auto

Sofia has recurrently expressed dissatisfaction with merge_auto not
working super well at reproducing data due to the update in some data
types not being fit for data science. For example, while Counter's
update does what one would expect, dict's update might simply override
previous values from the iterator without giving out any feedback. To
allow data scientists to avoid this issue, we add a new API that makes
sure that previous values don't exist.

berkeman commented:
Yes, a silent override is not optimal.  I wonder if we need a new function for this, or if it is sufficient to use the new "safe"-argument?  Also, talking to Carl, we should probably profile this to find the fastest way to check for existing keys (maybe using set.union()?).

pabloyoyoista commented:
We've met today and agreed that we want the new behavior to be the default, after some small testing showed that the new behavior only introduces an overhead of 1 second per 31 million rows. Of course, this breaks tests. So I'll fix them and we have a meeting queued in 2 weeks

pabloyoyoista committed:
[test_dataset_fanout] set hashlabel of previous dataset

pabloyoyoista committed:
[dataset_fanout_collect] allow_overwrite in merge_auto

We are basically returning a set, so we don't really care if the key
already exists
